File: ruby/139.word-break.rb
# @param {String} s
# @param {String[]} word_dict
# @return {Boolean}
def word_break(s, word_dict)
  return false if word_dict.empty?

  dp = Array.new(s.size + 1, false)
  dp[0] = true

  (1..s.size).each do |i|
    word_dict.select { |word| word.size <= i }.each do |word|
      dp[i] ||= dp[i - word.size] && s[i - word.size...i] == word
    end
  end

  dp.last
end

File: rust/src/139.word-break.rs
use crate::Solution;

impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        if word_dict.is_empty() {
            return false;
        }

        let mut dp = vec![false; s.len() + 1];
        dp[0] = true;

        for i in 1..=s.len() {
            for word in &word_dict {
                let word_len = word.len();
                if word_len <= i && dp[i - word_len] && &s[i - word_len..i] == word {
                    dp[i] = true;
                    break;
                }
            }
        }
        *dp.last().unwrap_or(&false)
    }
}
